### DISCUSSION TITLE
How to keep a ssh session alive?

### DISCUSSION THREAD

**Discussion by sla-te** [amoffat/sh#705] [Q&A] [answered]
I take it, implementing the suggestions from 2015 via https://github.com/amoffat/sh/issues/223, were still not considered?
Now due to me using this quite often, like for instance here: https://github.com/amoffat/sh/discussions/701#discussioncomment-7186549
Now I wonder if it is a good idea to use sh as backend, because the actual chaining is not really passing through any checks anymore and are being executed "raw", what are your thoughts?

**Comment by ecederstrand** [comment]
It's not clear from the linked issues exactly what your problem is, or what you are trying to achieve?

The syntax suggested in #223 is not supported by sh.

**Reply by sla-te** [reply]
Essentially I want to depict doing something like the following onto shs API and I am unsure how to do it properly: 

(bash)
```bash
touch my_file && while [ -f myfile ]; do echo "myfile still exists"; sleep 1; done && sleep 1
```

Why am I doing this?
I am facing the issue, that sh runs every Command in a new subprocess separately, which is not desirable when sending multiple commands via ssh, for instance:

```python
my_server = sh.ssh.bake("amoffat@10.10.10.100")
print(my_server("ifconfig"))  # <- new ssh session
print(my_server("whoami")) # <- new ssh session
# ... etc
```

I circumvented this behaviour via https://github.com/amoffat/sh/discussions/701#discussioncomment-7186549, which uses sshs [ControlMaster](https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Multiplexing) feature. It opens a "master" connection on a socket and keeps it opened as long as the "master" session is still alive.
Doing this significantly speeds up the process if running a lot of commands in succession on a remote server.

Now If I do
```python
import sh
master = sh.ssh.bake("-o", "ControlMaster=yes", "-o", "ControlPath=~/.ssh/control-10.0.0.1_root", "-i", "priv_key", "root@10.0.0.1")
```
I need a way to keep the master session alive, in order for me to use it to send commands as slave:
```python
import sh
slave = sh.ssh.bake("-o", "ControlPath=~/.ssh/control-10.0.0.1_root", "-i", "priv_key", "root@10.0.0.1")
print(slave("ifconfig"))  # <- using the "master" ssh session/socket
print(slave("whoami")) # <- using the "master" ssh session/socket
```

**Reply by amoffat** [reply]
Can you be more specific about how needing to keep the master session alive requires chaining and piping?

**Reply by sla-te** [reply]
Well, when I do `touch my_file && while [ -f myfile ]; do echo "myfile still exists"; sleep 1; done && sleep 1` I am, to my terms, chaining together `touch my_file` with `while [ -f myfile ...]`. -   But I do see your point, the topic is not really a perfect fit for the actual question, as of course `&&`does nothing else but asserting the previous command was run successfully, which could also be achieved by running both commands separately and checking the exit codes.

The only reason I "have to" do this is, that I could not think of a different way to keep the session alive but chaining a while loop after a command, that I can pass to sh for execution, so it will keep the session alive until I delete `myfile` via separate command (using the slave).

So I am actually not really sure which topic would fit the question better, maybe "How to keep an empty ssh session alive"? :D

**Reply by amoffat** [reply]
This doesn't work?

```python
master = sh.ssh("-o", "ControlMaster=yes", "-o", "ControlPath=~/.ssh/control-10.0.0.1_root", "-i", "priv_key", "root@10.0.0.1", _bg=True)
# ... do slave stuff ...
master.kill()
```

**Reply by sla-te** [reply]
Nope
```python
master = sh.ssh(
    "-o",
    "ControlMaster=yes",
    "-o",
    f"ControlPath=~/.ssh/control-10.0.0.1_root_ses_debugging",
    "-i",
    "priv_key",
    "root@10.0.0.1",
    "-p",
    "22",
    _bg=True,
)

while True:
    print(master.is_alive())
    time.sleep(1) 

>>> True
>>> False
>>> False
>>> False
>>> False
...
```
This behaviour seems logical to to me though as the command does not provide any work to be done, so sh would connect to the server and disconnect right after.

**Reply by amoffat** [reply, marked as answer]
Take a look at the "-N" option on ssh. It is for ssh sessions to stay alive when they don't execute a remote command.

**Reply by sla-te** [reply]
Genius. :) didn't figure that option could exist in ssh!

**Reply by amoffat** [reply]
Glad you got it working.

For future discussions, please try to make an extra effort to ensure that the issue is sh-related, and not related to your specific work that you're doing. Consider this discussion title "How to keep a ssh session alive" and its answer have nothing to do with sh. By the discussions you've made so far, it seems you are building some automated ssh-based system for your job or a personal project. That's great, but we can't help with that.

Sh is pretty much feature complete (likely nothing new will be added) and pretty mature (very few bugs). If you find yourself struggling with using it, it's probably related to the nuances of the programs you are launching, and not how sh itself is working. We'd like to keep discussions here related to issues with sh itself.

If you're struggling with other problems, ChatGPT is your friend, and sh has been around long enough that it has knowledge around its usage. As mentioned in a previous discussion, stackoverflow can be helpful for general questions too.

**Reply by sla-te** [reply]
I completely agree with you, that questions should be sh-related but at the same time I believe, that the use case to run multiple commands on a remote server is one is a very common one. Using this question in combination with the context manager I provided in my other question it will could motivate other developers to use sh over other libraries.

I really did my best to consume all the docs and references provided but it would sure have been nice to have some more examples on how to actually implement in a real world scenario. I had found the neat way to use ssh via using `bake` but if you open one isolated session for each command it does not really get close to the real world implementation, where it is very likely, that you dont just do `ip a` but rather want to run a little shell script wrapped in python. - For instance you want to cd to a certain folder set a variable in the shell, such as `SOME_LIB_PATH=\some\location\on\the\remote server`  and use that in the command you are calling, e.g. something like this: `VOLTA_HOME=$HOME/.volta && PATH=$VOLTA_HOME/bin:$PATH && cd /var/www/frontend && npm run pm2:production`. This was by the way one motivation for me even going down the path to try and persist the ssh session for follow up commands, which I now realized by using the above in combination with  tmux (which got pretty complex) - If I would have had examples about different strategies to achieve the desired result, eventually, this complexity would not even be required.

This library is the most clean one of everything I found but it lacks real world examples, which makes it really hard to use. If we had that it would be perfect (btw the internal subprocess handling is quite impressive).
